<!DOCTYPE html>
<html>
  <head>
    <title>Metagenomics with R</title>
    <meta charset="utf-8">
    <meta name="author" content="Antoine Bichat" />
    <meta name="date" content="2017-11-23" />
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Metagenomics with R
## R Users Grenoble
### Antoine Bichat
### November 23, 2017

---

class: center, middle, inverse




#Metagenomics

---

# What is this?

* Metagenomics is the study of the genomes of all species living in a given environment
.footnote[
[*] This presentation is simplified to make the topic understable in 5 minutes.
]

--

* One wants to know the composition in micro-organisms of different samples

--

* Principally _Bacteria_, _Archea_ and _Fungi_

--

* More and more companies are interested in metagenomics
.center[
![](Nestle-Research.jpg)
![](enterome.jpg)
]


---

# Abundance table

* Output from bioinformatic pipeline

* Input for statistical analysis

&lt;br&gt;

&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; Taxon &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; Sample1 &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; Sample2 &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; Sample3 &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Escherichia coli &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 27.8 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 22.1 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 19.0 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Enterobacter cloacae &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 23.8 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 16.5 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 24.2 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Bifidobacterium longum &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 7.9 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 21.2 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 27.9 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Klebsiella sp &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.6 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 10.5 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 11.5 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Staphylococcus aureus &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1.7 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 13.4 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 9.0 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Bacteroidetes fragilis &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 19.3 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 0.0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 0.8 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Other &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 15.9 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 16.3 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 7.6 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

---
class: middle, inverse, center

# R Pipeline

---
# Biological process 

* One gene (16S) present in all bacteria with variation is isolated and sequenced

&lt;br&gt;

--

* Output of the sequencing: One FASTA file per sample
```
&gt; GTCGATCGATGCCCTAGCCGATAGATCCCGATATAGCCGATAGAAAATATACGA...
&gt; GTCGATCGATGCCCTAGCCGATAGATCGCGATATAGCCGATAGAAAATATACGT...
&gt; GTCGATCGATGCCCTAGCCGATAGATCGCGATATAGCCGATAGAAAATATACGA...
&gt; GTCGATCGATGCCATAGCCGATAGATCCCGATATAGCCGATAGAAAATATACGA...
...
```
---
# Clustering

Similar sequences are grouped

* Similarity threshold in genomes

* Error correction of sequences
  
&lt;br&gt;

```
&gt; GTCGATCGATGCCCTAGCCGATAGATCCCGATATAGCCGATAGAAAATATACGA... -&gt; Group 1
&gt; GTCGATCGATGCCCTAGCCGATAGATCGCGATATAGCCGATAGAAAATATACGT... -&gt; Group 2
&gt; GTCGATCGATGCCCTAGCCGATAGATCGCGATATAGCCGATAGAAAATATACGA... -&gt; Group 1
&gt; GTCGATCGATGCCATAGCCGATAGATCCCGATATAGCCGATAGAAAATATACGA... -&gt; Group 3 
...
```
---
# Annotation 

A species is assignated to each group

* NBC (Naive Bayesian Classifier)
  
* BLAST (Basic Local Alignment Search Tool)

&lt;br&gt;

```
Group 1 -&gt; Escherichia coli
Group 2 -&gt; Enterobacter cloacae
Group 3 -&gt; Bifidobacterium longum
...
```
---
# DADA2

The R package [**dada2**](https://benjjneb.github.io/dada2/) uses error correction and NBC


&lt;br&gt;
&lt;br&gt;
&lt;br&gt;
&lt;br&gt;
.center[
&lt;img src="bioconductor_logo.jpg" alt="Drawing" style="width: 500px;"/&gt;
]


---
# Other bioinformatic tools
&lt;br&gt;

&lt;blockquote class="twitter-tweet" data-lang="fr" align="center"&gt;&lt;p lang="en" dir="ltr"&gt;The R package implementing our new method for identifying contaminants in amplicon/metagenomics data is available now, w/ documentation and tutorial vignette: &lt;a href="https://t.co/lL1PsF6HnB"&gt;https://t.co/lL1PsF6HnB&lt;/a&gt; &lt;a href="https://twitter.com/hashtag/kitome?src=hash&amp;amp;ref_src=twsrc%5Etfw"&gt;#kitome&lt;/a&gt; &lt;a href="https://twitter.com/hashtag/rstats?src=hash&amp;amp;ref_src=twsrc%5Etfw"&gt;#rstats&lt;/a&gt;&lt;/p&gt;&amp;mdash; Benjamin Callahan (@bejcal) &lt;a href="https://twitter.com/bejcal/status/932619941622239232?ref_src=twsrc%5Etfw"&gt;20 novembre 2017&lt;/a&gt;&lt;/blockquote&gt;

---
class: middle, center, inverse

# Data visualisation
## With [**ggplot2**](http://ggplot2.tidyverse.org/)

---
# Sample composition...
&lt;div style="margin-top: -30px"&gt;&lt;/div&gt;
according to delivery mode 

.center[
&lt;img src="barplot_compo.png" alt="Drawing" style="width: 650px;"/&gt;
]

---
# Repartition of groups ...
&lt;div style="margin-top: -30px"&gt;&lt;/div&gt;
according to delivery mode and age
&lt;br&gt;
&lt;br&gt;
Each sample is assignated to one group according to its composition

.center[
&lt;img src="barplot_groups.png" alt="Drawing" style="width: 600px;"/&gt;
]

---

class: center, middle, inverse

# Thanks for your attention :) 


.footnote[Slides created via the R package [**xaringan**](https://github.com/yihui/xaringan).]
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {window.dispatchEvent(new Event('resize'));});
(function() {var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler"); if (!r) return; s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }"; d.head.appendChild(s);})();</script>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
